File: lwk_bindings/tests/bindings/drain_lbtc.py
from lwk import *

# Start nodes
node = LwkTestEnv()

network = Network.regtest_default()
policy_asset = network.policy_asset()
client = ElectrumClient.from_url(node.electrum_url())

signer = Signer(Mnemonic.from_random(12), network)
wollet = Wollet(network, signer.wpkh_slip77_descriptor(), datadir=None)

# Fund both wallets
sats = 100_000
txid = node.send_to_address(wollet.address(0).address(), sats, asset=None)
wollet.wait_for_tx(txid, client)

node_addr = node.get_new_address()
# ANCHOR: drain_lbtc_wallet
# Create a PSET sending all LBTC to the node address
builder = network.tx_builder()
builder.drain_lbtc_wallet()
builder.drain_lbtc_to(node_addr)
pset = builder.finish(wollet)
# ANCHOR_END: drain_lbtc_wallet

pset = signer.sign(pset)
pset = wollet.finalize(pset)
tx = pset.extract_tx()
txid = client.broadcast(tx)
wollet.wait_for_tx(txid, client)

assert wollet.balance().get(policy_asset, 0) == 0
assert len(wollet.transactions()) == 2

File: lwk_bindings/go/drain_lbtc.go
package main

import (
	"log"

	"lwk-go/lwk"
)

func main() {
	// Start nodes
	node := lwk.NewLwkTestEnv()

	network := node.Network()
	policyAsset := network.PolicyAsset()
	client, err := lwk.ElectrumClientFromUrl(node.ElectrumUrl())
	if err != nil {
		log.Fatal(err)
	}

	signer, err := lwk.SignerRandom(network)
	if err != nil {
		log.Fatal(err)
	}
	desc, err := signer.WpkhSlip77Descriptor()
	if err != nil {
		log.Fatal(err)
	}
	wollet, err := lwk.NewWollet(network, desc, nil)
	if err != nil {
		log.Fatal(err)
	}

	// Fund both wallets (script says both, but only sets up one)
	sats := uint64(100_000)

	idx := uint32(0)
	addrResult, err := wollet.Address(&idx)
	if err != nil {
		log.Fatal(err)
	}

	txid := node.SendToAddress(addrResult.Address(), sats, nil)
	_, err = wollet.WaitForTx(txid, client)
	if err != nil {
		log.Fatal(err)
	}

	nodeAddr := node.GetNewAddress()

	// ANCHOR: drain_lbtc_wallet
	// Create a PSET sending all LBTC to the node address
	builder := network.TxBuilder()
	if err := builder.DrainLbtcWallet(); err != nil {
		log.Fatal(err)
	}
	if err := builder.DrainLbtcTo(nodeAddr); err != nil {
		log.Fatal(err)
	}
	pset, err := builder.Finish(wollet)
	if err != nil {
		log.Fatal(err)
	}
	// ANCHOR_END: drain_lbtc_wallet

	pset, err = signer.Sign(pset)
	if err != nil {
		log.Fatal(err)
	}

	tx, err := pset.Finalize()
	if err != nil {
		log.Fatal(err)
	}

	txid2, err := client.Broadcast(tx)
	if err != nil {
		log.Fatal(err)
	}

	_, err = wollet.WaitForTx(txid2, client)
	if err != nil {
		log.Fatal(err)
	}

	balance, err := wollet.Balance()
	if err != nil {
		log.Fatal(err)
	}
	if balance[policyAsset] != 0 {
		panic("expected 0 lbtc in wollet after drain")
	}

	txs, err := wollet.Transactions()
	if err != nil {
		log.Fatal(err)
	}
	if len(txs) != 2 {
		panic("expected exactly 2 transactions in wollet")
	}
}
